File: 算法/代码随想录算法PDF/00调试源代码/108.将有序数组转换为二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=108 lang=cpp
 *
 * [108] 将有序数组转换为二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if (nums.size() == 0) return nullptr;  // 特殊情况

        TreeNode* root = new TreeNode();  // 初始化
        queue<TreeNode*> nodeQ;
        queue<int> leftQ, rightQ;
        nodeQ.push(root);
        leftQ.push(0);
        rightQ.push(nums.size());

        while (!nodeQ.empty()) {
            TreeNode* node = nodeQ.front(); nodeQ.pop();  // 根节点赋值
            int left = leftQ.front(); leftQ.pop();
            int right = rightQ.front(); rightQ.pop();
            int mid = left + (right - left) / 2;
            node->val = nums[mid];

            if (left < mid) {  // 处理左区间
                node->left = new TreeNode();
                nodeQ.push(node->left);
                leftQ.push(left);
                rightQ.push(mid);
            }

            if (right > mid + 1) {  // 处理右区间
                node->right = new TreeNode();
                nodeQ.push(node->right);
                leftQ.push(mid + 1);
                rightQ.push(right);
            }
        }
        return root;
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/108.将有序数组转换为二叉搜索树.py
#
# @lc app=leetcode.cn id=108 lang=python3
#
# [108] 将有序数组转换为二叉搜索树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        if len(nums) == 0: return None
        root = TreeNode()
        nodeSt = [root]
        leftSt = [0]
        rightSt = [len(nums)]

        while nodeSt:
            node = nodeSt.pop()
            left = leftSt.pop()
            right = rightSt.pop()
            mid = left + (right - left) // 2
            node.val = nums[mid]

            if left < mid:
                node.left = TreeNode()
                nodeSt.append(node.left)
                leftSt.append(left)
                rightSt.append(mid)
            
            if right > mid + 1:
                node.right = TreeNode()
                nodeSt.append(node.right)
                leftSt.append(mid + 1)
                rightSt.append(right)
        
        return root
# @lc code=end

